### DISCUSSION TITLE
Intellisense not working?

### DISCUSSION THREAD

**Discussion by sparkwenc** [microsoft/vscode-cpptools#12624] [Q&A] [answered]
I am trying to configure c++23 intellisense on MacOS using g++-14 (rather than clang).

The hover works well, but the auto-completion, go-to-definition, etc., simply does not work.

By pressing Shift-Spc, its pops up with "No suggestions."

Is this something wrong with my configuration with intellisense or unsupported behavior for c++23?

**Comment by sean-mcmanus** [comment]
C++23 with g++-14 is supported on Mac. Can you set C_Cpp.loggingLevel to "Debug" and see if there are any errors reported? Does it work with C++20 or with clang with C++23? Do you have any sample code?

**Reply by sparkwenc** [reply]
I set the C_Cpp.loggingLevel to "Debug" and there are no errors reported, and my whole settings.json file is attached.

Up to now, I've fixed auto-completion, but go-to-definition still doesn't work, it pops up with "no definition found for '[anything I clicked]'".

I think it once worked with c++20 previously, but I was using the clang-command-adapter extension previously, and didn't manage to configure it for the intellisense provided by c/c++ extension.

```json
{
    "workbench.startupEditor": "none",
    "workbench.editorAssociations": {
        "*.ipynb": "jupyter-notebook"
    },
    "workbench.colorTheme": "Visual Studio Light - C++",
    "window.zoomLevel": 0,
    "git.autofetch": true,
    "git.enableSmartCommit": true,
    "gitlens.defaultDateStyle": "absolute",
    "explorer.confirmDelete": false,
    "editor.formatOnSave": true,
    "editor.minimap.enabled": true,
    "editor.renderWhitespace": "all",
    "editor.renderControlCharacters": true,
    "editor.fontFamily": "Monaco",
    "editor.fontSize": 16,
    "editor.insertSpaces": false,
    "editor.suggestSelection": "first",
    "editor.autoClosingBrackets": "languageDefined",
    "editor.autoClosingComments": "languageDefined",
    "editor.autoClosingQuotes": "languageDefined",
    "vsintellicode.modify.editor.suggestSelection": "automaticallyOverrodeDefaultValue",
    "breadcrumbs.enabled": true,
    "explorer.confirmDragAndDrop": false,
    "http.proxySupport": "off",
    "python.languageServer": "Default",
    "[python]": {
        "editor.formatOnType": true,
        "editor.defaultFormatter": "ms-python.autopep8"
    },
    "debug.onTaskErrors": "abort",
    "[cpp]": {
        "editor.formatOnType": true,
        "editor.defaultFormatter": "ms-vscode.cpptools"
    },
    "C_Cpp.default.cStandard": "c23",
    "C_Cpp.default.intelliSenseMode": "macos-gcc-x64",
    "C_Cpp.default.cppStandard": "c++23",
    "C_Cpp.default.compilerPath": "/usr/local/bin/g++-14",
    "C_Cpp.default.compilerArgs": [
        "-std=c++23",
        "-Wall",
        "-Wextra",
        "-Wconversion",
        "-O3"
    ],
    "C_Cpp.autocomplete": "default",
    "C_Cpp.autocompleteAddParentheses": true,
    "C_Cpp.clang_format_path": "/usr/local/Cellar/clang-format/18.1.8/bin/clang-format",
    "C_Cpp.clang_format_fallbackStyle": "LLVM",
    "C_Cpp.clang_format_sortIncludes": true,
    "C_Cpp.intelliSenseEngine": "default",
    "C_Cpp.intelliSenseEngineFallback": "enabled",
    "C_Cpp.codeAnalysis.clangTidy.enabled": false,
    "C_Cpp.loggingLevel": "Debug"
}
```

**Reply by sean-mcmanus** [reply]
@sparkwenc When you run C/C++: Log Diagnostics with the source file active does it show the expected configuration being used? 

Hover uses IntelliSense too. So I'm not sure why that would work and the other features not.

Completion doesn't work when you use `::`?

**Reply by sparkwenc** [reply]
Yes, the Log Diagnostics shows all the expected configurations I used.

I find it weird as well, completion and hover both work fine now, including when I use `::`. Only go-to-definition does not work currently.

Perhaps, is there any explicit command to reset this feature?

**Reply by sean-mcmanus** [reply]
@sparkwenc Go to definition could fail if there's a problem with the database, particularly if the definition is in a different TU (source file). You can try running C/C++: Reset IntelliSense Database to fix it (and wait for the file with the definition to be tag parsed).

You may also want to check the Outline pane to see if the definitions appear there or running the Go to Symbol in Workspace command -- if those don't find the definition, then the symbol is missing from the database for some reason (unrelated to IntelliSense parsing).

**Reply by sparkwenc** [reply]
Indeed! I run C/C++: Reset IntelliSense Database, and it doesn't fix it. I also checked the Outline pane and tried Go to Symbol in Workspace. They do not have any code symbols contained in my cpp code.

This seems to be a bit similar to the issue reported here [11001](https://github.com/microsoft/vscode-cpptools/issues/11001), but I haven't found a solution for it.

**Reply by sean-mcmanus** [reply, marked as answer]
@sparkwenc Do all symbols not appear at all in any workspace? There may be some failure to create the database, so you might try changing C_Cpp.default.browse.databaseFilename to something like "${workspacefolder}/.vscode/vc.db" to ensure the database is able to be created. Otherwise, if you have C_Cpp.loggingLevel set to "Debug" then you might want to check for tag parsing error messages after saving/editing the file with no symbols.

**Reply by sparkwenc** [reply]
Thank you!!!
I changed C_Cpp.default.browse.databaseFilename from empty to ${default} then the problem is fixed.

**Reply by sean-mcmanus** [reply]
@sparkwenc Oh, that is unexpected. I don't think `${default}` is a valid setting value, so I'm not sure what path is being used. But it sounds like there might have been a permissions issue with the original default path or database file.

**Reply by sparkwenc** [reply]
I don't know the value of ${default} either, the description of C_Cpp.default.browse.databaseFilename is
"The value to use in a configuration if browse.databaseFilename is either not specified or set to ${default}."

In practice it is writing what is needed into ${workspacefolder}/.vscode/settings.json -- is this expected?

**Reply by bobbrow** [reply]
Maybe we need to clarify the setting documentation.  It's trying to say that the value of `C_Cpp.default.browse.databaseFilename` is used if you set the value of `browse.databaseFilename` to `${default}` in your `c_cpp_properties.json` file.  Like this:

**c_cpp_properties.json**
```json
{
    "configurations": [
        {
            "name": "foo",
            "browse": {
               "databaseFilename": "${default}"
            }
        }
    ]
}
```

**Reply by sean-mcmanus** [reply]
Yeah, the description could add text "in c_cpp_properties.json" clarify. Ideally, some warning would also be exposed somehow -- I don't see any in our logging. I'd have to debug it to see what path is actually being used.
